dtfs-central-api/api-tests/v1/utilisation-reports/get-utilisation-report.api-test.js

```diff
@@ -1,47 +1,26 @@
-const wipeDB = require('../../wipeDB');
 const app = require('../../../src/createApp');
 const api = require('../../api')(app);
-const db = require('../../../src/drivers/db-client').default;
-const { DB_COLLECTIONS } = require('../../../src/constants/db-collections');
-const { MOCK_UTILISATION_REPORT } = require('../../mocks/utilisation-reports/utilisation-reports');
+const { MOCK_BANKS } = require('../../mocks/banks');
 
-const getUrl = (_id) => `/v1/utilisation-reports/${_id}`;
-
-describe('/v1/utilisation-reports/:_id', () => {
-  beforeAll(async () => {
-    await wipeDB.wipe([DB_COLLECTIONS.UTILISATION_REPORTS]);
-  });
+const getUrl = (id) => `/v1/utilisation-reports/${id}`;
 
+describe('/v1/utilisation-reports/:id', () => {
   describe('GET /v1/utilisation-reports/:_id', () => {
-    it('returns 400 when an invalid report MongoDB ID is provided', async () => {
+    it('returns 400 when an invalid report ID is provided', async () => {
       // Act
-      const { status, body } = await api.get(getUrl('invalid-mongo-id'));
+      const { status, body } = await api.get(getUrl('invalid-id'));
 
       // Assert
       expect(status).toEqual(400);
-      expect(body.errors.at(0).msg).toEqual("Invalid MongoDB '_id' path param provided");
+      expect(body.errors.at(0).msg).toEqual("Invalid 'id' path param provided");
     });
 
     it('gets a utilisation report', async () => {
-      // Arrange
-      const collection = await db.getCollection(DB_COLLECTIONS.UTILISATION_REPORTS);
-      const { insertedId } = await collection.insertOne(MOCK_UTILISATION_REPORT);
-      const _id = insertedId.toString();
-      const dateUploadedAsISOString = MOCK_UTILISATION_REPORT.dateUploaded.toISOString();
-
       // Act
-      const { body, status } = await api.get(getUrl(_id));
+      const { status } = await api.get(getUrl(MOCK_BANKS.BARCLAYS.id));
```

Comment:
I think this could be as simple as mocking out all the repo methods for all api-tests so that they all resolve but don't actually run (probably in the `api-test-setup.jest.config.js` file). This would also meet our criteria of wanting to limit the scope of api-tests

Reply:
I think there are a few options here, lots of good thoughts Josh! Sorry if this is irrelevant as I've only read the x lines of new code in this file and nothing else on this review yet!

My hot takes regarding databases and tests

Data in api-tests should stand alone -- ie it should not matter what happens before or after running an api test. To the point that we should definitely not care about `npm run load` at all, or whether or not any test has previously edited the database.

This pattern is used in a few places, notably around user creation. We create the data in the database in (or before) the test we use it. We revert (or delete) these changes afterwards. After all, in a dream world, we shouldn't need to do this, as each api test should start on an empty database. We actually clear the database in a lot of user based api tests.

In terms of mocking repo methods -- I'd argue its simply less time to insert the data into a clean database. We do this elsewhere on the project. It also means we don't have to worry about "what does this complex query need to be mocked with" and allows us to have a more end to end api test.

In terms of _how_ we insert this data in the database there are two ways -- one is by api-calls, the other is by direct insert statements.

I did not like api-calls at the start of this project, however, considering the complex way we modify data it allows us to have an accurate representation of what the data in the database will actually be like. In terms of testing pyramid, this should be sufficient for api testing mostly. I think they work on this project well for this reason. A downside is that we are also somewhat "testing" these endpoints we use to populate dummy data.

The second way, direct inserts, is what I would chose on a relational database (so could be useful here), but is somewhat at risk of falling out of date on non relational databases. For this -- I usually like to create a builder function so that these can be easily populated in the database from a base configuration (immer might be a good shout for this to avoid heavy builder functions). This can be useful for relational databases, but has the downside of us needing to keep these up-to-date if our schemas change (or in the case of non-relational databases, if the B.E changes how these are saved to the db -- and this can cause our tests to invisibly get out of sync). 

I think the two choices above warrant a wider discussion. I'd prefer the second on a new project with a relational database, but I'm aware existing patterns etc.

Thank you for coming to my todd talk. Very sorry if this is tangential

Reply:
This may be worth scheduling a meeting for offline as it would be good to get input from everyone I think. Does that sound like something we can do @ChristianMcc ?

Reply:
Generally I'm very much in the camp of preferring a fresh db for each integration tests which is obviously not the existing pattern and not something to embark on in this ticket. 

Josh did point out some of the existing tests rely on seeding so given that was going to be simplest I went for that pattern, although I did fail to remember we might have reports not uploaded as I am still trying to get my head around all the domain 😅 Happy to embark on setting up adding SQL inserts for the api tests, what sort of discussion is generally had on topics like this and should I be waiting for that to happen to work on this further? I'm keen for us to try and avoid wasted effort.